src/utils/graph.ml

```diff
@@ -0,0 +1,306 @@
+    (VSet.remove v vs
+    , ESet.filter (fun e -> not (is_vertex_in_edge e v)) es)
+           
+  (** Remove a [vertex] from a graph and its associated [edges] *)                             
+
+  let remove_edge:  t ->  edge ->  t
+    = fun (vs, es) e -> (vs, ESet.remove e es) 
+  (** Remove an [edge] from a graph *)                             
+
+  let non_target_vertices: t -> vertices
+    = fun (vs, es) ->
+    VSet.filter (fun v -> ESet.for_all (fun e -> not (is_vertex_target e v)) es) vs
+  (** Returns a list of all vertices that have no incoming edge  *)
+
+  let non_source_vertices: t -> vertices
+    = fun (vs, es) ->
+    VSet.filter (fun v -> ESet.for_all (fun e -> not (is_vertex_target e v)) es) vs
+  (** Returns a list of all vertices that have no outgoing edges  *)
+    
+  let connect: t -> vertex -> t = fun g v ->
+    let vs = get_vertices g in
+    (VSet.add v vs
+    , VSet.fold (fun v' es' -> ESet.add (mk_edge v v') es')
+        (get_vertices g)
+        (get_edges g))
+  (** Connect [vertex] that is already in the graph to all the other vertices in the graph *)
+
+
+  let sub_graph: t -> vertices -> t = fun g vs ->
+    ( vs
+    , ESet.filter (fun (src, tgt) -> VSet.mem src vs && VSet.mem tgt vs)
+        (get_edges g))
+  (** Gets a subgraph of given graph from a given set of vertices *)
+                                          
+  let is_point_graph: t -> bool = fun (vs, es) ->
+    ESet.is_empty es
+    
+  let union: t -> t -> t = fun (v1s, e1s) (v2s, e2s) ->
+    (VSet.union v1s v2s, ESet.union e1s e2s) 
+  (** Unions two graphs *)
+    
+  let reverse_topological_sort: t -> vertex list = fun ((vs, es) as g) ->
+    let rec r_topological_sort_helper: t -> vertex list -> vertex list
+      = fun ((vs, es) as g) sorted_vs ->
+      let no_outgoing_vs = non_source_vertices g in
+
+      Log.log L_trace
+        "-----------\nGraph state:\n %a\nSorted vertices: %a\n new non source vertices: %a\n-------------"	
+        pp_print_graph g	
+        (Lib.pp_print_list pp_print_vertex ",") sorted_vs	
+        pp_print_vertices no_outgoing_vs ;
+      (** graph is empty case *)
+      if VSet.is_empty no_outgoing_vs
+      then if not (is_empty g)
+           then raise (CyclicGraphException
+                         (List.map (Lib.string_of_t pp_print_vertex) (VSet.elements vs)))
+           else sorted_vs
+      else
+        let new_g = VSet.fold (fun v g' -> remove_vertex g' v) no_outgoing_vs g in
+        r_topological_sort_helper new_g (sorted_vs @ VSet.elements no_outgoing_vs)
+    in r_topological_sort_helper g []
+  (** Computes a topological ordering of vertices 
+   *  or throws an [CyclicGraphException] if the graph is cyclic.
+   *  Implimentation is based on Kahn's algorithm 
+   * https://en.wikipedia.org/wiki/Topological_sorting *)
+
+  let topological_sort: t -> vertex list = fun g ->
+    Log.log L_trace "Phase 1.1 Building graph\n---------\n"
+    ; reverse_topological_sort g |> List.rev     
```

Comment:
`reverse_topological_sort` seems to return the elements already in topological order. Why are you reversing the result?

Reply:
It is necessary given the way we build the dependency graph. 
`a -> b` means that `a` depends on `b`, the topological sort would mean `[a, b]`
but we need a dependency sort, not topological sort. Hence we expect an order `[b, a]`

Maybe I should change the name of the function...

Reply:
The current names of the functions are misleading. Since you are reversing the list explicitly before returning the result, that is, you are not returning the reversed list as as a side-effect of your implementation, you should consider reversing the list outside this function, in the function that computes dependencies.

Reply:
I have changed the function names now.